include/multipass/recursive_dir_iterator.h

```diff
@@ -249,6 +250,48 @@ class RecursiveDirIterator
     fs::recursive_directory_iterator iter;
     DirectoryEntry current;
 };
+
+// wrapper class around std::filesystem::directory_iterator used for mocking purposes
+class DirIterator
+{
+public:
+    DirIterator() = default;
+    DirIterator(const fs::path& path, std::error_code& err) : self{path / "."}, parent{path / ".."}, iter{path, err}
+    {
+    }
+
+    virtual bool hasNext()
+    {
+        return iter != fs::end(iter);
+    }
+
+    virtual const DirectoryEntry& next()
+    {
+        if (self)
+        {
+            const fs::directory_entry entry{*self};
+            self.reset();
+            return current = entry;
+        }
+
+        if (parent)
+        {
+            const fs::directory_entry entry{*parent};
+            parent.reset();
+            return current = entry;
+        }
+
+        return current = *iter++;
+    }
+
+    virtual ~DirIterator() = default;
+
+private:
+    std::optional<fs::path> self;
+    std::optional<fs::path> parent;
+    fs::directory_iterator iter;
+    DirectoryEntry current;
```

Comment:
Nice wrapper utility class, besides that, I have a few small comments.
1. About the data member `current`, it holds the same information as `*iter`. I think you did this double bookkeeping because you need to return `DirectoryEntry` reference. I am wondering if it is gotta be better if you return by value and remove all the `current` related code.
2. There are some implicit constructions from `fs::directory_entry` to `DirectoryEntry`, maybe it is better to change the constructor `DirectoryEntry(const fs::directory_entry& _entry)` to explicit and explicitly convert these.

Reply:
You are right, `current` is necessary for returning a `DirectoryEntry` reference. But the reason it returns a reference instead of a value is so that it can return a `MockDirectoryEntry` reference when writing unit tests.
As for the explicit constructors, sure, I will convert them.